### PR TITLE
Backport perl: add .withPackages #59215 #60028

### PR DESCRIPTION
**Pull request by samueldr**
Backports  #59215 and #60028 to 19.03

Undraft and merge only once #60028 has been validated and merged, please.

The initial commit (perl: add .withPackages) had two conflicts. The first one is a possibly missing perl update to stable (?) and the second one was with the signature where I think a cross-compilation change changed it (trivial to fix).

cc @volth

### CODE REVIEW

File: pkgs/development/interpreters/perl/default.nix [hunk 1]
```diff
@@ -179,6 +193,7 @@ let
 in rec {
   # the latest Maint version
   perl528 = common {
+    self = perl528;
```
Comment (volth) [inline pkgs/development/interpreters/perl/default.nix]: ```perldevel``` few lines below should have ```self = perldevel``` added too

perhaps there are more problems, this one was easy to spot

Comment (volth) [reply pkgs/development/interpreters/perl/default.nix]: ofborg failed here too :)

**Review by aanderse — APPROVED** [review]
After adding the missing `self = perldevel;` reference @volth mentioned, building my system against this branch, and then testing a few select perl scripts against an `environment.systemPackages = [ perl.withPackages... ];` (both `perl` and `perldevel`), I feel pretty good that this is working as expected without any obvious regressions.

**Comment by aanderse** [comment]
@samueldr I'm leaving final approval and merging to you for this change.

**Pull request merged**